tests/phpunit/integration/Core/Util/Entity_FactoryTest.php

```diff
@@ -0,0 +1,784 @@
+		// URL is not part of the site.
+		$this->assertNull( Entity_Factory::from_url( 'https://www.google.com' ) );
+
+		// Set 'show_on_front' to 'posts' home page archive.
+		update_option( 'show_on_front', 'posts' );
+
+		// Expect home 'blog' entity for home URL (and the above 'show_on_front').
+		$this->assertEntity(
+			new Entity(
+				user_trailingslashit( home_url() ),
+				array(
+					'type'  => 'blog',
+					'title' => 'Home',
+				)
+			),
+			Entity_Factory::from_url( home_url() )
+		);
+	}
+
+	/**
+	 * @dataProvider data_from_wp_query
+	 *
+	 * @param array       $query_args      Query arguments.
+	 * @param Entity|null $expected_entity Expected entity, or null.
+	 */
+	public function test_from_wp_query( array $query_args, Entity $expected_entity = null ) {
+		global $wp_rewrite;
+
+		add_filter(
+			'option_home',
+			function() {
+				return 'https://example.com';
+			}
+		);
+
+		// Run the actual query.
+		$query = new \WP_Query();
+		$query->query( $query_args );
+
+		// For date-based archive, set a fake post as query result because these archives only exist when there is
+		// content available for them.
+		if ( $expected_entity && in_array( $expected_entity->get_type(), array( 'year', 'month', 'day' ), true ) ) {
+			$year     = isset( $query_args['year'] ) ? $query_args['year'] : current_time( 'Y' );
+			$monthnum = isset( $query_args['monthnum'] ) ? $query_args['monthnum'] : current_time( 'm' );
+			$day      = isset( $query_args['day'] ) ? $query_args['day'] : current_time( 'd' );
+
+			$query->posts               = array( new \WP_Post( new \stdClass() ) );
+			$query->posts[0]->post_date = "{$year}-{$monthnum}-{$day} 12:00:00";
+			$query->posts[0]->filter    = 'raw'; // Avoids attempt to fetch this fake post.
+		}
+
+		$entity = Entity_Factory::from_wp_query( $query );
+		if ( $expected_entity ) {
+			// Fill in entity IDs because they are unknown to the data provider.
+			if ( in_array( $expected_entity->get_type(), array( 'post', 'blog' ), true ) && ! $expected_entity->get_id() ) {
+				$title = $this->extract_wp_entity_title_from_url_entity_title( $expected_entity->get_title() );
+				if ( isset( self::$post_titles_to_ids[ $title ] ) ) {
+					$this->force_set_property( $expected_entity, 'id', (int) self::$post_titles_to_ids[ $title ] );
+				}
+			} elseif ( 'term' === $expected_entity->get_type() ) {
+				$title = $this->extract_wp_entity_title_from_url_entity_title( $expected_entity->get_title() );
+				if ( isset( self::$term_names_to_ids[ $title ] ) ) {
+					$this->force_set_property( $expected_entity, 'id', (int) self::$term_names_to_ids[ $title ] );
+				}
+			} elseif ( 'user' === $expected_entity->get_type() ) {
+				$title = $this->extract_wp_entity_title_from_url_entity_title( $expected_entity->get_title() );
+				if ( isset( self::$user_display_names_to_ids[ $title ] ) ) {
+					$this->force_set_property( $expected_entity, 'id', (int) self::$user_display_names_to_ids[ $title ] );
+				}
+			}
```

Comment:
This seems unnecessary - does the data provider not have access to `self::$post_titles_to_ids` and others to set the ID in this way? If so, let's elaborate on the comment to explain why they're unknown to the provider.

Reply:
I found the reason why in [the docs](https://phpunit.readthedocs.io/en/7.0/writing-tests-for-phpunit.html#data-providers):
> All data providers are executed before both the call to the `setUpBeforeClass()` static method and the first call to the `setUp()` method. Because of that you can’t access any variables you create there from within a data provider. This is required in order for PHPUnit to be able to compute the total number of tests.

Reply:
After taking a closer look, we can avoid the need to mutate the `$expected_entity` in the test by returning a function from the data provider which returns the expected entity instead of returning it directly from the provider.
Then we only need to invoke the function in the test to get the expected entity; at which point it will have access to the full context of the `TestCase` so referencing the ID of the entity that was created during the set up is no problem.

E.g. this works:
(simplified)
```php
public function test_from_wp_query( array $query_args, \Closure $get_expected_entity ) {
	$expected_entity = $get_expected_entity();

	// ...

	// Run the actual query.
	$query = new \WP_Query();
	$query->query( $query_args );

	// ...

	$entity = Entity_Factory::from_wp_query( $query );

	if ( $expected_entity ) {
		$this->assertEntity( $expected_entity, $entity );
	} else {
		$this->assertNull( $entity );
	}
}

public function data_from_wp_query() {
	return array(
		'front page'                   => array(
			array(),
			function () {
				// The full context of the TestCase will be available when this is called.
				return new Entity(
					'https://example.com/',
					array(
						'type'  => 'post',
						'title' => 'Home',
						'id'    => self::$post_titles_to_ids['Home'],
					)
				);
			},
		),
		// ...
	);
}
```

I think we should opt for this approach to avoid complicated rewriting and keep the test method as simple as possible. It makes the provider method a little more complicated/verbose, but I think the tradeoff is definitely worth it.
Keep in mind that we could also use the `factory` to create a WP entity in the test from a provider-provided function like this too so they wouldn't have to all be created beforehand. Since the same enitites are referenced between other tests in this test case, I think it's okay to share them, just highlighting the other options and flexibility this has.